### PR TITLE
Initial work

### PR DESCRIPTION
**Pull request by jotak**
Replaces #1 

Initial move from vertx-hawkular-metrics to global vertx-monitoring

Based on PR vert-x3/vertx-hawkular-metrics#65 from @tsegismont and @dankristensen

Includes only influxdb reporter at the moment

### CODE REVIEW

**Comment by jotak** [comment]
@vietj or @tsegismont can you review please?

**Comment by vietj** [comment]
I've assigned Thomas formally to this, I'll throw an eye though to the PR.

File: batching-collector/pom.xml [hunk 1]
```diff
@@ -26,7 +26,7 @@
 
   <artifactId>vertx-monitoring-batching-collector</artifactId>
 
-  <name>Vert.x Metrics Collector</name>
+  <name>Vert.x Monitoring Batching Collector</name>
```
[the remark this answers is not in the document]
Comment (tsegismont) [reply batching-collector/pom.xml]: In fact, why split the project into _common_ and _collector_?

There's just one collector implementation, it's not meant to be generic. In the original PR (https://github.com/vert-x3/vertx-hawkular-metrics/pull/65) there was just a _collector_ project

File: batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+import java.util.Map;
+
+import static io.vertx.ext.monitoring.common.MetricsCategory.*;
+
+/**
+ * Metrics SPI implementation.
+ *
+ * @author Thomas Segismont
+ */
+public abstract class BatchingVertxMetrics<T extends BatchingReporterOptions> extends DummyVertxMetrics {
```
Comment (tsegismont) [inline batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java]: See the other comment: in the original PR, this class was named `VertxMetricsBase`. It is not a _batching_ Vert.x Metrics impl. It's the base HTTP reporter class which is _batching_: it sends metrics in batches to the remote server.

File: batching-collector/src/main/java/io/vertx/ext/monitoring/collector/BatchingReporterOptions.java [hunk 3]
```diff
@@ -26,27 +26,27 @@
...
  */
 @DataObject(generateConverter = true, inheritConverter = true)
-public abstract class BatchingReporterOptions extends MetricsOptionsBase {
+public class BatchingReporterOptions extends MetricsOptionsBase {
 
   /**
    * The default metric name prefix (empty).
    */
-  public static final String DEFAULT_PREFIX = "";
+  private static final String DEFAULT_PREFIX = "";
```
Comment (tsegismont) [inline batching-collector/src/main/java/io/vertx/ext/monitoring/collector/BatchingReporterOptions.java]: Default values are always exposed in Vert.x options classes

File: batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/Scheduler.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Collects metrics and relay them to the sender.
+ *
+ * @author Thomas Segismont
+ */
+public class Scheduler {
+  private final Vertx vertx;
+  private final Handler<List<DataPoint>> sender;
```
Comment (tsegismont) [inline batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/Scheduler.java]: Now that there's a `Reporter` interface, we should probably use this instead of the generic `Handler<List<DataPoint>>`, wdyt?

File: reporters/influxdb/README.adoc [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+This project is an InfluxDb implementation of the Vert.x Metrics SPI: metrics built from Vert.x events will be sent to
+an InfluxDb Timeseries database.
+
+Please see the in-source Asciidoc documentation or the main documentation on the web-site for a full description
+of this component:
+
+* link:http://vertx.io/docs/vertx-monitoring-influxdb/[web-site docs]
+* link:src/main/asciidoc[in-source docs]
+
+If you want to test you can use this docker command to run an InfluxDb instance:
```
Comment (tsegismont) [inline reporters/influxdb/README.adoc]: Could you rephrase this? It seems to imply you need this to run the Influx reporter test while we're actually mocking the Influx server

File: reporters/influxdb/src/test/java/io/vertx/ext/monitoring/influxdb/impl/InfluxDbReporterITest.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  public void shouldSendDataToInfluxDb(TestContext context) {
+    Async async = context.async();
+    vertx.createHttpServer(new HttpServerOptions()
+      .setCompressionSupported(true)
+      .setDecompressionSupported(true)
+      .setLogActivity(true)
+      .setHost("localhost")
+      .setPort(8086))
+      .requestHandler(req -> {
+        req.exceptionHandler(t -> System.out.println("An exception occured handling request: " + t));
```
Comment (tsegismont) [inline reporters/influxdb/src/test/java/io/vertx/ext/monitoring/influxdb/impl/InfluxDbReporterITest.java]: Fail the `context`?

File: batching-collector/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jotak) [reply batching-collector/pom.xml]: This is because I have a Prometheus reporter ready for a next PR, and it will be using common but not collector. So it's already prepared for split.

[the remark this answers is not in the document]
Comment (jotak) [reply batching-collector/pom.xml]: About why _batching_: 
In the previous PR HttpReporterBase was defined in collector and was setting down the mechanism for sending batches to reporters (it still does). So I wanted to make it clear that this "collector" module is actually designed to send batches to the reporter.

However I don't mind to change back to _collector_ to make it simpler.

File: batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java (hunk 2 above)
Comment (jotak) [reply batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java]: `VertxMetricsBase` is too generic while this will only be used to send batches with a scheduler. Again, Prometheus won't use it. So I had to find something less generic than just _XXBase_.

**Comment by vietj** [comment]
before going further, the package should be `io.vertx.monitoring` and not `io.vertx.ext.monitoring`

**Comment by vietj** [comment]
the module collector module shall simply be named `vertx-monitoring-collector` (no need to say it's batching)

the module directories should be named:

- vertx-monitoring-collector
- vertx-monitoring-common
- vertx-monitoring-reporters

File: batching-collector/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tsegismont) [reply batching-collector/pom.xml]: I would prefer to see all the code in this PR, to get a better picture. Is that ok @vietj

[the remark this answers is not in the document]
Comment (tsegismont) [reply batching-collector/pom.xml]: >In the previous PR HttpReporterBase was defined in collector and was setting down the mechanism for sending batches to reporters (it still does)

Yes, `HttpReporterBase`. Not `VertxMetricsBase`. So the collector is not a _batching_. It's only the base reporter class

[the remark this answers is not in the document]
Comment (vietj) [reply batching-collector/pom.xml]: I agree with @tsegismont I also wondered where the Prometheus support was. An initial-work PR should be complete.

File: batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java (hunk 2 above)
Comment (tsegismont) [reply batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java]: I need to see the code for Prometheus. It sounds like we're going to maintain multiple collector implementations. We wanted to avoid that.

File: batching-collector/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jotak) [reply batching-collector/pom.xml]: It was clear for me that in a first time I would take in just InfluxDB (that's in the PR head comment). I've never been aware than prometheus was expected for 3.5.1
Btw "completeness" depends on what you expect. It's 100% ready for use.

File: batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java (hunk 2 above)
Comment (jotak) [reply batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java]: You can see it here [1], although it's not yet adapted to the new module "vertx-monitoring", it's roughly what I plan to put in.

[1] https://github.com/jotak/vertx-metrics/tree/prom/reporters/prometheus/src/main/java/io/vertx/ext/prometheus/impl

And yes it doesn't fit in this collector implementation. I talked about it previously (don't remember where and when but I can search). This is mainly because of the pull-based model. This is also because by using the prometheus registry (from the prometheus java client), we gain some functionalities that we don't have otherwise (collecting summaries, histograms, building timers out of summaries, etc.). I think there's too much to lose if we absolutely wanted to do something very generic.

Comment (jotak) [reply batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java]: Note also that some metrics colelcted for influx are pointless to collect for prometheus. Like eventbus' `receivedMessages` / `receivedLocalMessages` / `receivedRemoteMessages`. In prometheus it's just 1 metric with a label.

File: batching-collector/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tsegismont) [reply batching-collector/pom.xml]: > It was clear for me that in a first time I would take in just InfluxDB (that's in the PR head comment)

I noticed your comment. I asked to include the Prometheus code as it seems we will not be reusing the collector implementation in the end.

>I've never been aware than prometheus was expected for 3.5.1

Me neither. But it seems to impact the way this project will work so I think it's better to consider what has been done already.

**Comment by jotak** [comment]
Feedbacks done
I also added prometheus.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/README.adoc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+* link:http://vertx.io/docs/vertx-monitoring-prometheus/[web-site docs]
+* link:src/main/asciidoc[in-source docs]
+
+== Quick setup
+
+To use the Prometheus reporter, you need first to get a running Prometheus instance.
+See link:https://prometheus.io/docs/prometheus/latest/installation/[Prometheus installation].
+
+Configure Prometheus to scrape your Vert.X server. Example in `prometheus.yml`:
```
Comment (tsegismont) [inline vertx-monitoring-reporters/vertx-monitoring-prometheus/README.adoc]: `Vert.x` not `Vert.X`

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/README.adoc [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+```yml
+scrape_configs:
+  - job_name: 'myapp'
+    static_configs:
+      - targets: ['localhost:8080']
```
Comment (tsegismont) [inline vertx-monitoring-reporters/vertx-monitoring-prometheus/README.adoc]: Use another port as many users will have http servers running on this port

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/pom.xml [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+      <version>${prometheus-client.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>io.vertx</groupId>
+      <artifactId>vertx-monitoring-common</artifactId>
+      <version>3.5.1-SNAPSHOT</version>
+    </dependency>
+    <dependency>
+      <groupId>io.vertx</groupId>
+      <artifactId>vertx-web</artifactId>
```
Comment (tsegismont) [inline vertx-monitoring-reporters/vertx-monitoring-prometheus/pom.xml]: We cannot require to add Vert.x Web for monitoring with Prometheus

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+----
+
+== Configuration
+
+=== Using an embedded HTTP server for exposing endpoints
+
+[source,groovy]
+----
+Code not translatable
```
Comment (tsegismont) [inline vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc]: Update the example so that it can be translated to Groovy

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+----
+def vertx = Vertx.vertx([
+  metricsOptions:[
+    enabled:true
+  ]
+])
+
+// Later on, creating a router
+def router = Router.router(vertx)
+router.route("/custom").handler(io.vertx.monitoring.prometheus.impl.PrometheusVertxMetrics.createMetricsHandler())
```
Comment (tsegismont) [inline vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc]: It is nice to be able to plug a handler into an existing Vert.x Web app, avoiding to create bind multiple ports. But the reporter should be able to start a Vert.x core server if needed

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/pom.xml (hunk 9 above)
Comment (tsegismont) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/pom.xml]: > I forgot to mark it optional. Would it be ok if it's optional?

Yes

> The reason for this is to allow "embedded prometheus server" to start automatically without the user having to manually create a webserver/router etc. for prometheus.

The reporter should start a Vert.x core HttpServer without requiring anything from the user else than some config.

>It's configurable in options, and if the user doesn't activate the embedded server then he doesn't have to provide that dependency.

How will the metrics get collected then?

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc (hunk 11 above)
Comment (jotak) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc]: It's the purpose of the above example, the one you cannot see because "code is not translatable" :)

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+|===
+|Metric type
+|Metric name
+|Description
+
+|Gauge
+|`vertx_verticle{name=<name>}`
+|Number of verticle instances deployed.
+
+|===
```
[the remark this answers is not in the document]
Comment (jotak) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc]: Yeah it's a difference indeed, and it's somewhat related to issue #2 , basically at the moment the best way to have application-defined metrics with prometheus is to use the shared registry directly (just similar to the dropwizard impl).
I think this point could be addressed during the tech-preview phase, to make sure we provide a long term solution for this.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+  public VertxPrometheusOptions setEnabled(boolean enable) {
+    super.setEnabled(enable);
+    return this;
+  }
+
+  /**
+   * An embedded server will start to expose metrics with Prometheus format
+   * @param serverOptions the server options
+   */
+  public VertxPrometheusOptions embedServer(VertxPrometheusServerOptions serverOptions) {
```
Comment (tsegismont) [inline vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java]: Does this works well with codegen? Usually options classes have get/is,set methods.

Comment (tsegismont) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java]: It would be better to have an `enabled` field in `VertxPrometheusServerOptions`

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusServerOptions.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * Vert.x Prometheus embedded server configuration.
+ *
+ * @author Joel Takvorian
+ */
+@DataObject(generateConverter = true, inheritConverter = true)
+public class VertxPrometheusServerOptions {
```
Comment (tsegismont) [inline vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusServerOptions.java]: Can you add an `HttpServerOptions` field? Users could then configure the endpoint (ssl in particular)

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/pom.xml (hunk 9 above)
Comment (jotak) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/pom.xml]: > The reporter should start a Vert.x core HttpServer without requiring anything from the user else than some config.

It's the case

> How will the metrics get collected then?

Without embedded server, the metrics are still collected in the registry, and that registry is accessible to the user so that he can do whatever he wants. It's interesting in the case where you want to reuse an existing router, as shown in one of the examples.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java (hunk 13 above)
Comment (jotak) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java]: I'm changing this, maybe it's the reason why the groovy example is not generated

About the `enabled` field, I've seen it's used in a lot of option classes, but to be honest I'm a bit perplexed about it. If I set Options XYZ, why would I want to have it disabled?
In other words I would prefer to write: `new VertxPrometheusOptions().setEmbeddedServerOptions(new EmbeddedServerOptions());`, where it's implictly enabled, rather than 
`new VertxPrometheusOptions().setEmbeddedServerOptions(new EmbeddedServerOptions().setEnabled());` which adds verbosity for nothing.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusServerOptions.java (hunk 14 above)
Comment (jotak) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusServerOptions.java]: Actually, I'm replacing `VertxPrometheusServerOptions` with `HttpServerOptions`.
I'm adding the `String endpoint` field directly in `VertxPrometheusOptions`

Doing so is much easier to use, no need to encapsulate options in options in options.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/dataobjects.adoc [hunk 15]
```diff
@@ -36,24 +44,12 @@ Set whether metrics will be enabled on the Vert.x instance. Metrics are not enab
...
-[frame="topbot"]
-|===
-^|Name | Type ^| Description
-|[[endpoint]]`endpoint`|`String`|-
-|[[host]]`host`|`String`|-
-|[[port]]`port`|`Number (int)`|-
+|[[registryName]]`registryName`|`String`|
++++
+Set a name for the prometheus registry, so that a new registry will be created and associated with this name.
+ To retrieve this registry later, call link
```
Comment (jotak) [inline vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/dataobjects.adoc]: It seems that doc generation doesn't like all links...

File: vertx-monitoring-reporters/vertx-monitoring-influxdb/src/main/java/io/vertx/monitoring/influxdb/package-info.java [hunk 16]
```diff
@@ -0,0 +1,424 @@
...
+ * If the metric is a counter or availability, or if you prefer explicit configuration, set the {@code type} and/or {@code timestamp} attributes:
+ *
+ * [source,$lang]
+ * ----
+ * {@link examples.MetricsExamples#userDefinedMetricExplicit()}
+ * ----
+ *
+ * NOTE: InfluxDb understands all timestamps as milliseconds since January 1, 1970, 00:00:00 UTC.
+ */
+@ModuleGen(name = "vertx-influxdb", groupPackage = "io.vertx")
```
Comment (jotak) [inline vertx-monitoring-reporters/vertx-monitoring-influxdb/src/main/java/io/vertx/monitoring/influxdb/package-info.java]: @tsegismont I'm not sure if "ModuleGen" must take name of the artifact, or if it's not necessary?
Here the maven artifact is vertx-monitoring-influxdb

**Comment by jotak** [comment]
Did some small updates as requested by @vietj (upgrade parent pom...). Also fixed doc gen for "links" in dataobjects.adoc.

File: batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java (hunk 2 above)
Comment (tsegismont) [reply batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java]: @jotak yes I remember we talked about the problem but not about a solution. Or maybe you did with @vietj ? The thing is that having if we have two collector and reporters then there's no point in creating this module with reporter and collectors subprojects.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/pom.xml (hunk 9 above)
Comment (tsegismont) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/pom.xml]: Thanks for the details

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc (hunk 11 above)
Comment (tsegismont) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc]: Can you investigate why it's not translatable?

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc (hunk 12 above)
[the remark this answers is not in the document]
Comment (tsegismont) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc]: Is the event bus bridged used by the Prometheus reporter? It seems it is as there related fields in the options class.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java (hunk 13 above)
Comment (tsegismont) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java]: Because Vert.x is polyglot and when you write JS or Ruby verticles, you don't even have option classes. Just JSON or Hashes.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc (hunk 11 above)
Comment (jotak) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc]: It's fixed now. I had to use only pojo-style getters/setters in the examples.

File: batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java (hunk 2 above)
Comment (jotak) [reply batching-collector/src/main/java/io/vertx/ext/monitoring/collector/impl/BatchingVertxMetrics.java]: Maybe @vietj can comment.
My understanding is that vertx-monitoring-parent will be the new parent for all metrics-related modules. For instance, dropwizard module could be move here later (but if we want to make it consistent we would have to rename it to smthg like vertx-monitoring-dropwizard, which is a breaking change, so for the time being we don't touch it)

But still, it's true that since hawkular is removed this architecture is now a bit weird with a generic-style collector that is used only for 1 reporter, influx. However it can be used later if a new push-based TSDB needs to be implemented here. So we may not want to destroy something that has been designed to handle various reporters. It's not as if we had a huge benefit of breaking it. My opinion is that we can keep it as is.

File: vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc (hunk 12 above)
[the remark this answers is not in the document]
Comment (jotak) [reply vertx-monitoring-reporters/vertx-monitoring-prometheus/src/main/asciidoc/groovy/index.adoc]: Removed from prometheus options

**Comment by tsegismont** [comment]
@jotak it won't be possible to resolve the duplication issue for 3.5.1. However, we want to release both reporters as a tech preview.

To move things forward, could you please move everything to a single `vertx-monitoring` Maven module ? This will avoid having many jars which may disappear in the next release. The Influx and Prometheus specific code can still be separated in different packages (two `package-info` files will produce two Asciidoc files). We'll then arrange the separation on the Vert.x website.

**Comment by jotak** [comment]
Just merged everything into a single module. The doc files are incorrect though (still need to be sorted out)

File: src/main/java/examples/MetricsExamples.java [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+import io.vertx.monitoring.influxdb.VertxInfluxDbOptions;
+import io.vertx.monitoring.prometheus.VertxPrometheusOptions;
+import io.vertx.monitoring.prometheus.impl.PrometheusVertxMetrics;
+
+/**
+ * @author Thomas Segismont
+ * @author Joel Takvorian
+ */
+@Source
+@SuppressWarnings("unused")
```
Comment (tsegismont) [inline src/main/java/examples/MetricsExamples.java]: Please remove this. You shouldn't have unused methods in this class, they should all be referred from the Influx reporter `package-info` file.

Note that some examples look like they were copied from the Hawkular ones but are not relevant to Influx

File: src/main/asciidoc/js/influxdb.adoc [hunk 18]
```diff
@@ -0,0 +1,419 @@
...
+|===
+
+== User defined metrics
+
+Users can send their own metrics to the InfluxDb server. In order to do so, the event bus metrics bridge must be
+enabled:
+
+[source,js]
+----
+Code not translatable
```
Comment (tsegismont) [inline src/main/asciidoc/js/influxdb.adoc]: Some examples are still not translated

File: src/main/java/io/vertx/monitoring/VertxMetricsFactoryImpl.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    } else {
+      return new InfluxDbVertxMetrics(vertx, new VertxInfluxDbOptions(metricsOptions.toJson()));
+    }
+  }
+
+  @Override
+  public MetricsOptions newOptions() {
+    // Tricky hack to guess if we're expecting Prometheus options or Influx
+    try {
+      Class.forName("io.prometheus.client.CollectorRegistry");
```
Comment (tsegismont) [inline src/main/java/io/vertx/monitoring/VertxMetricsFactoryImpl.java]: Can you replace with:

```java
Class.forName("io.prometheus.client.CollectorRegistry", false, getClass().getClassLoader())
```

Then the class will not be initialized until it's actually used (useful if metrics are disabled).

File: src/main/java/io/vertx/monitoring/prometheus/impl/PrometheusVertxMetrics.java [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+import static io.vertx.monitoring.common.MetricsCategory.*;
+
+/**
+ * Metrics SPI implementation for Prometheus.
+ *
+ * @author Joel Takvorian
+ */
+public class PrometheusVertxMetrics extends DummyVertxMetrics {
+  private final CollectorRegistry registry;
+  private final Optional<EventBusMetrics> eventBusMetrics;
```
Comment (tsegismont) [inline src/main/java/io/vertx/monitoring/prometheus/impl/PrometheusVertxMetrics.java]: Don't use `Optional` as a field. Its value is only used inside this class.

File: src/main/java/io/vertx/monitoring/prometheus/impl/PrometheusVertxMetrics.java [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+   * Create a metrics handler for vertx router, using the registry identified by {@code registryName}
+   */
+  public static Handler<RoutingContext> createMetricsHandler(String registryName) {
+    // Create handler from io.prometheus:simpleclient_vertx
+    return new MetricsHandler(PrometheusRegistries.get(registryName));
+  }
+
+  private HttpServer startServer() {
+    HttpServerOptions serverOptions = options.getEmbeddedServerOptions();
+    Router router = Router.router(vertx);
```
Comment (tsegismont) [inline src/main/java/io/vertx/monitoring/prometheus/impl/PrometheusVertxMetrics.java]: Vert.x Web is not really optional if it's needed even for the embedded server. Or am I missing something?

File: src/main/java/io/vertx/monitoring/prometheus/package-info.java [hunk 22]
```diff
@@ -0,0 +1,388 @@
...
+ * {@link io.vertx.core.eventbus.EventBus} and handlers
+ * * User defined metrics can be added to the shared registry (using Prometheus https://github.com/prometheus/client_java[simple Java client])
+ *
+ * == Prerequisites
+ *
+ * Follow the https://prometheus.io/docs/prometheus/latest/getting_started/[instructions to get Prometheus up and running].
+ *
+ * == Getting started
+ *
+ * The _${maven.artifactId}_ module must be present in the classpath.
```
Comment (tsegismont) [inline src/main/java/io/vertx/monitoring/prometheus/package-info.java]: Can you also say that the Prometheus client must be on the classpath?

Comment (tsegismont) [reply src/main/java/io/vertx/monitoring/prometheus/package-info.java]: BTW, why do we need this `simpleclient` JAR if we only expose metrics on an HTTP server?

File: src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+   * By default, disables <i>remote</i> label for net/http servers.
+   */
+  public static final boolean DEFAULT_ENABLE_REMOTE_LABEL_FOR_SERVERS = false;
+
+  /**
+   * The default metrics endpoint = /metrics when using an embedded server.
+   */
+  public static final String DEFAULT_EMBEDDED_SERVER_ENDPOINT = "/metrics";
+
+  private HttpServerOptions embeddedServerOptions;
```
Comment (tsegismont) [inline src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java]: Don't use null as an "enabled" indicator. This is not how we deal with options in general.

This field should be initialized with the default `HttpServerOptions` and then you can add an `startEmbeddedServer` boolean.

File: src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+
+  /**
+   * By default, uses the standard prometheus registry (name is null)
+   */
+  public static final String DEFAULT_REGISTRY_NAME = null;
+
+  /**
+   * By default, enables <i>remote</i> label for net/http clients.
+   */
+  public static final boolean DEFAULT_ENABLE_REMOTE_LABEL_FOR_CLIENTS = true;
```
Comment (tsegismont) [inline src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java]: 👍

File: src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+
+  /**
+   * By default, disables <i>remote</i> label for net/http servers.
+   */
+  public static final boolean DEFAULT_ENABLE_REMOTE_LABEL_FOR_SERVERS = false;
```
Comment (tsegismont) [inline src/main/java/io/vertx/monitoring/prometheus/VertxPrometheusOptions.java]: 👍

File: src/main/java/io/vertx/monitoring/prometheus/impl/Labels.java [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+ * limitations under the License.
+ */
+package io.vertx.monitoring.prometheus.impl;
+
+import io.vertx.core.net.SocketAddress;
+
+/**
+ * @author Joel Takvorian
+ */
+public final class Labels {
```
Comment (tsegismont) [inline src/main/java/io/vertx/monitoring/prometheus/impl/Labels.java]: Why not an enum?
